[BUG]: `nested_constraints` incompatible with `TemplateExpression`

### What happened?
I think conceptually these two options should work together.

What I want is for `log` to never be nested inside another `log` in `b(...)` or `e(...)`

```julia
atlas_template = TemplateStructure{(:b, :e)}(
     ((; b, e), (x,)) -> b((1-x))^e(x)
)

sr_model2 = SRRegressor(
    niterations=150,
    binary_operators=[+, -, *, ^],
    unary_operators=[log],
    should_optimize_constants=true,
    seed=nothing,
    nested_constraints = [log => [log => 0]],
    expression_type=TemplateExpression,
    expression_options=(; structure=atlas_template),
)

sr_mach2 = machine(sr_model2, X, ys)

fit!(sr_mach2)
```

### Version

1.1.0

### Operating System

Linux

### Interface

Jupyter Notebook

### Relevant log output

```shell
nested task error: TaskFailedException
    Stacktrace:
     [1] wait(t::Task)
       @ Base ./task.jl:370
     [2] fetch
       @ ./task.jl:390 [inlined]
     [3] (::SymbolicRegression.var"#56#61"{SymbolicRegression.SearchUtilsModule.SearchState{Float64, Float64, TemplateExpression{Float64, TemplateStructure{(:b, :e), var"#11#12", @NamedTuple{b::Int64, e::Int64}}, Node{Float64}, ComposableExpression{Float64, Node{Float64}, @NamedTuple{operators::DynamicExpressions.OperatorEnumModule.OperatorEnum{Tuple{typeof(+), typeof(-), typeof(*), typeof(safe_pow)}, Tuple{typeof(safe_log)}}, variable_names::Nothing}}, @NamedTuple{b::ComposableExpression{Float64, Node{Float64}, @NamedTuple{operators::DynamicExpressions.OperatorEnumModule.OperatorEnum{Tuple{typeof(+), typeof(-), typeof(*), typeof(safe_pow)}, Tuple{typeof(safe_log)}}, variable_names::Nothing}}, e::ComposableExpression{Float64, Node{Float64}, @NamedTuple{operators::DynamicExpressions.OperatorEnumModule.OperatorEnum{Tuple{typeof(+), typeof(-), typeof(*), typeof(safe_pow)}, Tuple{typeof(safe_log)}}, variable_names::Nothing}}}, @NamedTuple{structure::TemplateStructure{(:b, :e), var"#11#12", @NamedTuple{b::Int64, e::Int64}}, operators::DynamicExpressions.OperatorEnumModule.OperatorEnum{Tuple{typeof(+), typeof(-), typeof(*), typeof(safe_pow)}, Tuple{typeof(safe_log)}}, variable_names::Nothing}}, Task, Channel}, Int64, Int64})()
       @ SymbolicRegression ~/.julia/packages/SymbolicRegression/glgvH/src/SymbolicRegression.jl:816
    
        nested task error: AssertionError: options.should_simplify
        Stacktrace:
          [1] mutate!(tree::TemplateExpression{Float64, TemplateStructure{(:b, :e), var"#11#12", @NamedTuple{b::Int64, e::Int64}}, Node{Float64}, ComposableExpression{Float64, Node{Float64}, @NamedTuple{operators::DynamicExpressions.OperatorEnumModule.OperatorEnum{Tuple{typeof(+), typeof(-), typeof(*), typeof(safe_pow)}, Tuple{typeof(safe_log)}}, variable_names::Nothing}}, @NamedTuple{b::ComposableExpression{Float64, Node{Float64}, @NamedTuple{operators::DynamicExpressions.OperatorEnumModule.OperatorEnum{Tuple{typeof(+), typeof(-), typeof(*), typeof(safe_pow)}, Tuple{typeof(safe_log)}}, variable_names::Nothing}}, e::ComposableExpression{Float64, Node{Float64}, @NamedTuple{operators::DynamicExpressions.OperatorEnumModule.OperatorEnum{Tuple{typeof(+), typeof(-), typeof(*), typeof(safe_pow)}, Tuple{typeof(safe_log)}}, variable_names::Nothing}}}, @NamedTuple{structure::TemplateStructure{(:b, :e), var"#11#12", @NamedTuple{b::Int64, e::Int64}}, operators::DynamicExpressions.OperatorEnumModule.OperatorEnum{Tuple{typeof(+), typeof(-), typeof(*), typeof(safe_pow)}, Tuple{typeof(safe_log)}}, variable_names::Nothing}}, member::PopMember{Float64, Float64, TemplateExpression{Float64, TemplateStructure{(:b, :e), var"#11#12", @NamedTuple{b::Int64, e::Int64}}, Node{Float64}, ComposableExpression{Float64, Node{Float64}, @NamedTuple{operators::DynamicExpressions.OperatorEnumModule.OperatorEnum{Tuple{typeof(+), typeof(-), typeof(*), typeof(safe_pow)}, Tuple{typeof(safe_log)}}, variable_names::Nothing}}, @NamedTuple{b::ComposableExpression{Float64, Node{Float64}, @NamedTuple{operators::DynamicExpressions.OperatorEnumModule.OperatorEnum{Tuple{typeof(+), typeof(-), typeof(*), typeof(safe_pow)}, Tuple{typeof(safe_log)}}, variable_names::Nothing}}, e::ComposableExpression{Float64, Node{Float64}, @NamedTuple{operators::DynamicExpressions.OperatorEnumModule.OperatorEnum{Tuple{typeof(+), typeof(-), typeof(*), typeof(safe_pow)}, Tuple{typeof(safe_log)}}, variable_names::Nothing}}}, @NamedTuple{structure::TemplateStructure{(:b, :e), var"#11#12", @NamedTuple{b::Int64, e::Int64}}, operators::DynamicExpressions.OperatorEnumModule.OperatorEnum{Tuple{typeof(+), typeof(-), typeof(*), typeof(safe_pow)}, Tuple{typeof(safe_log)}}, variable_names::Nothing}}}, ::Val{:simplify}, ::MutationWeights, options::Options{SymbolicRegression.CoreModule.OptionsStructModule.ComplexityMapping{Int64, Int64}, DynamicExpressions.OperatorEnumModule.OperatorEnum{Tuple{typeof(+), typeof(-), typeof(*), typeof(safe_pow)}, Tuple{typeof(safe_log)}}, Node, TemplateExpression, @NamedTuple{structure::TemplateStructure{(:b, :e), var"#11#12", @NamedTuple{b::Int64, e::Int64}}}, MutationWeights, false, false, nothing, Nothing, 5}; recorder::Dict{String, Any}, parent_ref::Int64, kws::@Kwargs{temperature::Float64, dataset::Dataset{Float64, Float64, Matrix{Float64}, Vector{Float64}, Nothing, @NamedTuple{}, Nothing, Nothing, Nothing, Nothing}, score::Float64, loss::Float64, curmaxsize::Int64, nfeatures::Int64})
            @ SymbolicRegression.MutateModule ~/.julia/packages/SymbolicRegression/glgvH/src/Mutate.jl:555
          [2] mutate!
            @ ~/.julia/packages/SymbolicRegression/glgvH/src/Mutate.jl:545 [inlined]
          [3] macro expansion
            @ ~/.julia/packages/SymbolicRegression/glgvH/src/Mutate.jl:361 [inlined]
          [4] _dispatch_mutations!(tree::TemplateExpression{Float64, TemplateStructure{(:b, :e), var"#11#12", @NamedTuple{b::Int64, e::Int64}}, Node{Float64}, ComposableExpression{Float64, Node{Float64}, @NamedTuple{operators::DynamicExpressions.OperatorEnumModule.OperatorEnum{Tuple{typeof(+), typeof(-), typeof(*), typeof(safe_pow)}, Tuple{typeof(safe_log)}}, variable_names::Nothing}}, @NamedTuple{b::ComposableExpression{Float64, Node{Float64}, @NamedTuple{operators::DynamicExpressions.OperatorEnumModule.OperatorEnum{Tuple{typeof(+), typeof(-), typeof(*), typeof(safe_pow)}, Tuple{typeof(safe_log)}}, variable_names::Nothing}}, e::ComposableExpression{Float64, Node{Float64}, @NamedTuple{operators::DynamicExpressions.OperatorEnumModule.OperatorEnum{Tuple{typeof(+), typeof(-), typeof(*), typeof(safe_pow)}, Tuple{typeof(safe_log)}}, variable_names::Nothing}}}, @NamedTuple{structure::TemplateStructure{(:b, :e), var"#11#12", @NamedTuple{b::Int64, e::Int64}}, operators::DynamicExpressions.OperatorEnumModule.OperatorEnum{Tuple{typeof(+), typeof(-), typeof(*), typeof(safe_pow)}, Tuple{typeof(safe_log)}}, variable_names::Nothing}}, member::PopMember{Float64, Float64, TemplateExpression{Float64, TemplateStructure{(:b, :e), var"#11#12", @NamedTuple{b::Int64, e::Int64}}, Node{Float64}, ComposableExpression{Float64, Node{Float64}, @NamedTuple{operators::DynamicExpressions.OperatorEnumModule.OperatorEnum{Tuple{typeof(+), typeof(-), typeof(*), typeof(safe_pow)}, Tuple{typeof(safe_log)}}, variable_names::Nothing}}, @NamedTuple{b::ComposableExpression{Float64, Node{Float64}, @NamedTuple{operators::DynamicExpressions.OperatorEnumModule.OperatorEnum{Tuple{typeof(+), typeof(-), typeof(*), typeof(safe_pow)}, Tuple{typeof(safe_log)}}, variable_names::Nothing}}, e::ComposableExpression{Float64, Node{Float64}, @NamedTuple{operators::DynamicExpressions.OperatorEnumModule.OperatorEnum{Tuple{typeof(+), typeof(-), typeof(*), typeof(safe_pow)}, Tuple{typeof(safe_log)}}, variable_names::Nothing}}}, @NamedTuple{structure::TemplateStructure{(:b, :e), var"#11#12", @NamedTuple{b::Int64, e::Int64}}, operators::DynamicExpressions.OperatorEnumModule.OperatorEnum{Tuple{typeof(+), typeof(-), typeof(*), typeof(safe_pow)}, Tuple{typeof(safe_log)}}, variable_names::Nothing}}}, mutation_choice::Symbol, weights::MutationWeights, options::Options{SymbolicRegression.CoreModule.OptionsStructModule.ComplexityMapping{Int64, Int64}, DynamicExpressions.OperatorEnumModule.OperatorEnum{Tuple{typeof(+), typeof(-), typeof(*), typeof(safe_pow)}, Tuple{typeof(safe_log)}}, Node, TemplateExpression, @NamedTuple{structure::TemplateStructure{(:b, :e), var"#11#12", @NamedTuple{b::Int64, e::Int64}}}, MutationWeights, false, false, nothing, Nothing, 5}; kws::@Kwargs{recorder::Dict{String, Any}, temperature::Float64, dataset::Dataset{Float64, Float64, Matrix{Float64}, Vector{Float64}, Nothing, @NamedTuple{}, Nothing, Nothing, Nothing, Nothing}, score::Float64, loss::Float64, parent_ref::Int64, curmaxsize::Int64, nfeatures::Int64})
            @ SymbolicRegression.MutateModule ~/.julia/packages/SymbolicRegression/glgvH/src/Mutate.jl:346
          [5] _dispatch_mutations!
```


### Extra Info

_No response_

--- a/Project.toml
+++ b/Project.toml
@@ -1,7 +1,7 @@
 name = "SymbolicRegression"
 uuid = "8254be44-1295-4e6a-a16d-46603ac705cb"
 authors = ["MilesCranmer <miles.cranmer@gmail.com>"]
-version = "1.1.0"
+version = "1.1.1"
 
 [deps]
 ADTypes = "47edcb42-4c32-4615-8424-f2b9edc5f35b"

--- a/Project.toml
+++ b/Project.toml
@@ -1,7 +1,7 @@
 name = "SymbolicRegression"
 uuid = "8254be44-1295-4e6a-a16d-46603ac705cb"
 authors = ["MilesCranmer <miles.cranmer@gmail.com>"]
-version = "1.1.1"
+version = "1.2.0"
 
 [deps]
 ADTypes = "47edcb42-4c32-4615-8424-f2b9edc5f35b"

--- a/src/ComposableExpression.jl
+++ b/src/ComposableExpression.jl
@@ -106,6 +106,9 @@ for name in (:combine_operators, :simplify_tree!)
     end
 end
 
+function DE.count_scalar_constants(ex::AbstractComposableExpression)
+    return DE.count_scalar_constants(convert(Expression, ex))
+end
 function CO.count_constants_for_optimization(ex::AbstractComposableExpression)
     return CO.count_constants_for_optimization(convert(Expression, ex))
 end

--- a/src/TemplateExpression.jl
+++ b/src/TemplateExpression.jl
@@ -464,6 +464,9 @@ function DE.simplify_tree!(
     return with_contents(ex, new_contents)
 end
 
+function DE.count_scalar_constants(ex::TemplateExpression)
+    return sum(DE.count_scalar_constants, values(get_contents(ex)))
+end
 function CO.count_constants_for_optimization(ex::TemplateExpression)
     return sum(CO.count_constants_for_optimization, values(get_contents(ex)))
 end

--- a/src/TemplateExpression.jl
+++ b/src/TemplateExpression.jl
@@ -19,7 +19,8 @@ using DynamicExpressions:
     get_variable_names,
     get_tree,
     node_type,
-    count_nodes
+    count_nodes,
+    preserve_sharing
 using DynamicExpressions.InterfacesModule:
     ExpressionInterface, Interfaces, @implements, all_ei_methods_except, Arguments
 
@@ -372,7 +373,24 @@ function MM.condition_mutation_weights!(
     @nospecialize(options::AbstractOptions),
     curmaxsize::Int,
 ) where {T,L,N<:TemplateExpression,P<:PopMember{T,L,N}}
-    # HACK TODO
+    if !preserve_sharing(typeof(member.tree))
+        weights.form_connection = 0.0
+        weights.break_connection = 0.0
+    end
+
+    MM.condition_mutate_constant!(typeof(member.tree), weights, member, options, curmaxsize)
+
+    complexity = ComplexityModule.compute_complexity(member, options)
+
+    if complexity >= curmaxsize
+        # If equation is too big, don't add new operators
+        weights.add_node = 0.0
+        weights.insert_node = 0.0
+    end
+
+    if !options.should_simplify
+        weights.simplify = 0.0
+    end
     return nothing
 end
 

